presto-hive/src/main/java/com/facebook/presto/hive/S3SelectPushdown.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.metastore.Column;
+import com.facebook.presto.hive.metastore.Partition;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.log.Logger;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.compress.BZip2Codec;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.GzipCodec;
+import org.apache.hadoop.mapred.InputFormat;
+import org.apache.hadoop.mapred.TextInputFormat;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.facebook.presto.hive.HiveUtil.getCompressionCodec;
+import static com.facebook.presto.hive.HiveUtil.getDeserializerClassName;
+import static com.facebook.presto.hive.HiveUtil.getInputFormatName;
+import static com.facebook.presto.hive.metastore.MetastoreUtil.getHiveSchema;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_LOCATION;
+
+/**
+ * S3SelectPushdown uses Amazon S3 Select to push down queries to Amazon S3. This allows Presto to retrieve only a
+ * subset of data rather than retrieving the full S3 object thus improving Presto query performance.
+ */
+public class S3SelectPushdown
```

Comment:
In general I agree with you. I also don't like users to attach debuggers to figure out what Presto is doing, but I am also not a big fan of logging every decision in the code. In this class we have a log in every branch where we return `false`. I think you can just mention those cases in the documentation that s3 select will not be enabled/used for each of these cases instead of logging those. (You can say in the documentation (if you haven't already said) that only serdes X,Y,Z and file format T, U, V are supported, etc.) Also there are some obviously unnecessary logs, for example, on L143 and L148 if the config doesn't enable this feature then we don't need to log it, because it's obviously expected that the feature won't be enabled.